llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1393 @@
+	BasicBlockGroupData* parent;
+	std::deque<BasicBlockGroupData> subGroups;
+	std::map<const llvm::BasicBlock*, int> subGroupsID;
+	bool visitingAll{false};
+	int iterationCount{-1};
+	std::vector<std::pair<const llvm::BasicBlock*, int> > escaping;
+	static const DeterministicBBSet getAllBasicBlocks(const llvm::Function& F)
+	{
+		DeterministicBBSet set;
+		for (const llvm::BasicBlock& bb : F)
+		{
+			set.insert(&bb);
+		}
+		return set;
+	}
+	void splitIntoSCCs(std::deque<BasicBlockGroupData>& blockQueue, std::map<const llvm::BasicBlock*, int>& blockToIndexMap);
+public:
+	BasicBlockGroupData(FunctionData& data, BasicBlockGroupData* parentBBGD, const DeterministicBBSet& inner, const llvm::BasicBlock* start = nullptr)
+		: data(data), inner(inner), start(start), from(nullptr), parent(parentBBGD)
+	{
+		if (start)
+			assert(start->getParent() == data.getFunction());
+	}
+	BasicBlockGroupData(FunctionData& data)
+		: BasicBlockGroupData(data, /*parentBBGD*/nullptr, getAllBasicBlocks(*data.getFunction()), &data.getFunction()->getEntryBlock())
+	{
+	}
+	BasicBlockGroupData(BasicBlockGroupData& BBGData)
+		: data(BBGData.data), inner(BBGData.inner), start(nullptr), from(nullptr), parent(&BBGData)
+	{
+	}
+	void addIncomingEdge(const llvm::BasicBlock* comingFrom, const llvm::BasicBlock* target)
+	{
+		assert(inner.count(target));
+		if (start == nullptr)
+		{
+			start = target;
+			from = comingFrom;
+		}
+		if (start != target)
+		{
+			multiHead = true;
+			from = nullptr;
+		}
+		if (comingFrom != from)
+		{
+			from = nullptr;
+		}
+	}
+	void runVisitBasicBlock(const llvm::BasicBlock& BB, std::vector<const llvm::BasicBlock*>& visitNext)
+	{
+		assert(visitNext.empty());
+
+		data.currentEE->setIncomingBB(from);
+		const BasicBlock* ret = data.currentEE->visitBasicBlock<PartialInterpreter::VisitingPolicy::COMPUTE_VALUES>(&BB, from);
+
+		if (ret)
+		{
+			visitNext.push_back(ret);
+		}
+		else
+		{
+			llvm::DenseSet<const llvm::BasicBlock*> setSuccessors;
+			for (auto* bb : successors(&BB))
+			{
+				if (setSuccessors.insert(bb).second)
+					visitNext.push_back(bb);
+			}
+		}
+		//Do the visit of the BB, with comingFrom (possibly nullptr) as predecessor
+		//Loop backs will be directed to another BBgroup
+		//The visit will return the set of reachable BBs -> add them to the subGroups data accordingly
+	}
+	bool shouldCleanUp()
+	{
+		for (auto& p : escaping)
+		{
+			//llvm::errs() <<"escaping:\t" << p.first->getName() << "\t" << p.second <<"\n";
+			if (data.visitCounter[p.first] != p.second)
+				return true;
+		}
+		return false;
+	}
+	bool notifySuccessor(const llvm::BasicBlock* from, const llvm::BasicBlock* succ, const int iter)
+	{
+		if (inner.count(succ) == 0)
+		{
+			assert(parent);
+			//It should handled by the parent SCC
+
+			const bool notified = parent->notifySuccessor(from, succ, iter);
+
+			if (notified && !visitingAll)
+			{
+				escaping.push_back({from, iter});
+			}
+
+			return false;
+		}
+		else if (visitingAll)
```

Comment:
It looks like it would be better not to call notifysuccessor in the visitingAll case. It may be possible to remove the flag completely

Reply:
notifySuccessor has to be visited, because in the logic of the Tree of Subsets (https://github.com/leaningtech/cheerp-compiler/pull/64#discussion_r809416633) visitAll means "all basic blocks at the current level are considered visited" (and so you can skip the actual visit there), but notifySuccessors might still point to siblings somewhere else in the tree, and notify successor is the way to propagate the information. It could be simplifyicable by dropping visitngAll entierly, and probably it's just overall clearer.